lib/server/services/environment/EnvironmentService.js

```diff
@@ -117,6 +117,52 @@ class EnvironmentService {
         return this.get(environmentId);
     }
 
+    /**
+     * Define environment state at a given time
+     *
+     * @param {string} environmentId the id of the environment for which a new state must be registered
+     * @param {EnvironmentState} state the state to save
+     * @param {object<string, string>} vars updated environment vars
+     * @return {Promise<void>} resolves once the state has been saved
+     */
+    async createOrUpdateEnvironment(environmentId, state, vars) {
+        await dataSource.transaction(async () => {
+            const environment = await this.get(environmentId);
+
+            if (!environment) {
+                await this.create(
+                    {
+                        id: environmentId,
+                        createdAt: state.timestamp,
+                        rawConfiguration: JSON.stringify(vars),
+                    },
+                    state,
+                );
+            } else {
+                let configuration;
+                if (vars) {
+                    const { rawConfiguration } = environment;
+                    try {
+                        configuration = {
+                            ...rawConfiguration ? JSON.parse(rawConfiguration) : {},
+                            ...vars,
+                        };
+                    } catch (_) {
+                        // Do nothing
+                    }
+                }
+                await this.update(
+                    environmentId,
+                    {
+                        updatedAt: state.timestamp,
+                        rawConfiguration: JSON.stringify(configuration),
+                    },
+                    state,
+                );
```

Comment:
If for some reason the try-catch block fails, `configuration` remains `undefined`, what can happen only if `rawConfiguration` is not valid json, then you update invalid rawConfiguration with `undefined`. Is it intended?

Reply:
`JSON.stringify(undefined)` returns undefined